### PR TITLE
Am 840 delete api

### PR DESCRIPTION
**Pull request by dharmendra8186**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/AM-840

### Change description ###
Development of Delete role assignments by actor id or process/reference

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/roleassignment/controller/endpoints/DeleteAssignmentController.java [hunk 1]
```diff
@@ -2,12 +2,56 @@
...
+            code = 204,
+            message = "No Content"
+        ),
+        @ApiResponse(
+            code = 400,
+            message = V1.Error.INVALID_ROLE_NAME
+        ),
+        @ApiResponse(
+            code = 404,
+            message = V1.Error.ROLE_NOT_FOUND
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignment/controller/endpoints/DeleteAssignmentController.java:43]: It should be ROLE_ASSIGNMENT_NOT_FOUND

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/common/ParseRequestService.java [hunk 2]
```diff
@@ -77,4 +78,33 @@ private void setCorrelationId(Request request) throws Exception {
     public void removeCorrelationLog() throws Exception {
         correlationInterceptorUtil.afterCompletion();
     }
+
+    public Request prepareRequestObject(String process, String reference) throws Exception {
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/common/ParseRequestService.java:82]: rename it to prepareDeleteRequest

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/common/PersistenceService.java [hunk 3]
```diff
@@ -156,5 +156,20 @@ public void deleteRoleAssignment(RoleAssignment roleAssignment) {
         roleAssignmentRepository.delete(entity);
     }
 
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    public void deleteRoleAssignmentByActorId(UUID actorId) {
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/common/PersistenceService.java:160]: Rename: deleteAssignmentsByActorrId

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/common/PersistenceService.java [hunk 4]
```diff
@@ -156,5 +156,20 @@ public void deleteRoleAssignment(RoleAssignment roleAssignment) {
...
+
+        roleAssignmentRepository.deleteByActorId(actorId);
+    }
+
+    @Transactional(propagation = Propagation.REQUIRES_NEW)
+    public List<RequestedRole> getRoleAssignment(UUID actorId) {
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/common/PersistenceService.java:166]: Rename : getAssignmentsByActor

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java [hunk 5]
```diff
@@ -1,14 +1,177 @@
...
+        this.validationModelService = validationModelService;
+    }
+
+    public ResponseEntity<Object> deleteRoleAssignment(String actorId,
+                                                       String process,
+                                                       String reference) throws Exception {
+        List<RequestedRole> requestedRoles = null;
+
+        //1. create the request Object
+        Request request = parseRequestService.prepareRequestObject(process, reference);
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java:45]: what if process or reference are empty? we need to use this method only when both are available and throw 400 bad request if any one of them is not provided.

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java [hunk 6]
```diff
@@ -1,14 +1,177 @@
...
+        if (actorId != null) {
+            requestedRoles = persistenceService.getRoleAssignment(UUID.fromString(actorId));
+
+        } else if (process != null && reference != null) {
+            requestedRoles = persistenceService.getExistingRoleByProcessAndReference(
+                process,
+                reference,
+                Status.LIVE.toString()
+            );
+
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java:63]: There should be an else block to throw 400 bad request saying actorId or process+reference must be provided.

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java [hunk 7]
```diff
@@ -1,14 +1,177 @@
...
+        //calling drools rules for validation
+        validationModelService.validateRequest(assignmentRequest);
+
+
+    }
+
+    private void updateStatusAndPersist(Request request) {
+        for (RequestedRole requestedRole : assignmentRequest.getRequestedRoles()) {
+            requestedRole.setRequest(request);
+            if (!requestedRole.getStatus().equals(Status.APPROVED)) {
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java:95]: we need to set DELETE_APPROVED directly in Drools, if it is not DELETE_APPROVED then set it to DELETE_REJECTED

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java [hunk 8]
```diff
@@ -1,14 +1,177 @@
...
+
+            // Update request status to REJECTED
+            updateRequestStatus(parsedAssignmentRequest, Status.REJECTED);
+
+        }
+    }
+
+    private void deleteLiveRecords(AssignmentRequest assignmentRequest) {
+        for (RoleAssignment requestedRole : assignmentRequest.getRequestedRoles()) {
+            persistenceService.deleteRoleAssignmentByActorId(requestedRole.getActorId());
```
Comment (aashishaggarwal123) [inline src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java:144]: we can't delete by Actor if request is for process+reference

File: src/main/java/uk/gov/hmcts/reform/roleassignment/controller/endpoints/DeleteAssignmentController.java (hunk 1 above)
Comment (dharmendra8186) [reply src/main/java/uk/gov/hmcts/reform/roleassignment/controller/endpoints/DeleteAssignmentController.java:43]: Done

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/common/ParseRequestService.java (hunk 2 above)
Comment (dharmendra8186) [reply src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/common/ParseRequestService.java:82]: Done

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/common/PersistenceService.java (hunk 3 above)
Comment (dharmendra8186) [reply src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/common/PersistenceService.java:160]: Done.

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/common/PersistenceService.java (hunk 4 above)
Comment (dharmendra8186) [reply src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/common/PersistenceService.java:166]: Done.

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java (hunk 5 above)
Comment (dharmendra8186) [reply src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java:45]: Done.

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java (hunk 6 above)
Comment (dharmendra8186) [reply src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java:63]: Done.

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java (hunk 7 above)
Comment (dharmendra8186) [reply src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java:95]: Done.

File: src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java (hunk 8 above)
Comment (dharmendra8186) [reply src/main/java/uk/gov/hmcts/reform/roleassignment/domain/service/deleteroles/DeleteRoleAssignmentOrchestrator.java:144]: Done.

**Pull request merged**